File: src/mjorbit/cpu/orbit/propagator.py
"""RK4 chief orbit propagator.

Units: km, km/s, s.
"""

from __future__ import annotations

from typing import Optional
import numpy as np
from mjorbit.cpu.orbit.state import OrbitState
from mjorbit.cpu.orbit.gravity import total_accel


def _derivatives(
    R: np.ndarray,
    V: np.ndarray,
    use_j2: bool,
    a_external: Optional[np.ndarray],
) -> tuple[np.ndarray, np.ndarray]:
    """State derivatives: dR/dt = V, dV/dt = a(R) + a_external."""
    a = total_accel(R, use_j2=use_j2)
    if a_external is not None:
        a = a + a_external
    return V, a


def propagate_rk4(
    state: OrbitState,
    dt: float,
    use_j2: bool = True,
    a_external: Optional[np.ndarray] = None,
) -> OrbitState:
    """Propagate chief orbit by dt using RK4.

    Args:
        state: current chief orbit state
        dt: timestep in seconds
        use_j2: include J2 perturbation
        a_external: optional constant external acceleration in ECI, km/s^2
            (e.g. from thruster/drag orbit feedback)

    Returns:
        new OrbitState at t + dt
    """
    R, V = state.R_eci, state.V_eci

    k1R, k1V = _derivatives(R, V, use_j2, a_external)
    k2R, k2V = _derivatives(R + 0.5 * dt * k1R, V + 0.5 * dt * k1V, use_j2, a_external)
    k3R, k3V = _derivatives(R + 0.5 * dt * k2R, V + 0.5 * dt * k2V, use_j2, a_external)
    k4R, k4V = _derivatives(R + dt * k3R, V + dt * k3V, use_j2, a_external)

    R_new = R + (dt / 6.0) * (k1R + 2.0 * k2R + 2.0 * k3R + k4R)
    V_new = V + (dt / 6.0) * (k1V + 2.0 * k2V + 2.0 * k3V + k4V)

    return OrbitState(R_eci=R_new, V_eci=V_new, t=state.t + dt)

File: src/cpp/src/propagator.cc
#include "mujoco_orbit/propagator.h"

#include "mujoco_orbit/gravity.h"
#include "mujoco_orbit/math_utils.h"

namespace mujoco_orbit {
namespace {

void derivatives(
    const double R[3],
    const double V[3],
    bool use_j2,
    const double* a_external,
    double out_dR[3],
    double out_dV[3]) {
  detail::copy3(V, out_dR);
  total_accel(R, out_dV, use_j2);
  if (a_external) {
    detail::add3(out_dV, a_external, out_dV);
  }
}

}  // namespace

void propagate_rk4(
    const double R_eci[3],
    const double V_eci[3],
    double t,
    double dt,
    double out_R_eci[3],
    double out_V_eci[3],
    double* out_t,
    bool use_j2,
    const double* a_external) {
  double k1R[3];
  double k1V[3];
  derivatives(R_eci, V_eci, use_j2, a_external, k1R, k1V);

  double R_tmp[3];
  double V_tmp[3];

  detail::add_scaled3(R_eci, k1R, 0.5 * dt, R_tmp);
  detail::add_scaled3(V_eci, k1V, 0.5 * dt, V_tmp);
  double k2R[3];
  double k2V[3];
  derivatives(R_tmp, V_tmp, use_j2, a_external, k2R, k2V);

  detail::add_scaled3(R_eci, k2R, 0.5 * dt, R_tmp);
  detail::add_scaled3(V_eci, k2V, 0.5 * dt, V_tmp);
  double k3R[3];
  double k3V[3];
  derivatives(R_tmp, V_tmp, use_j2, a_external, k3R, k3V);

  detail::add_scaled3(R_eci, k3R, dt, R_tmp);
  detail::add_scaled3(V_eci, k3V, dt, V_tmp);
  double k4R[3];
  double k4V[3];
  derivatives(R_tmp, V_tmp, use_j2, a_external, k4R, k4V);

  for (int i = 0; i < 3; ++i) {
    out_R_eci[i] = R_eci[i] + (dt / 6.0) * (k1R[i] + 2.0 * k2R[i] + 2.0 * k3R[i] + k4R[i]);
    out_V_eci[i] = V_eci[i] + (dt / 6.0) * (k1V[i] + 2.0 * k2V[i] + 2.0 * k3V[i] + k4V[i]);
  }

  if (out_t) {
    *out_t = t + dt;
  }
}

}  // namespace mujoco_orbit
